git-webhook/pkg/webhook/webhook.go

```diff
@@ -0,0 +1,144 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"github.com/presslabs/wordpress-operator/pkg/apis/wordpress/v1alpha1"
+	"net/http"
+	"os"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"github.com/drone/go-scm/scm"
+	"github.com/drone/go-scm/scm/driver/github"
+	"github.com/go-logr/logr"
+)
+
+type Server struct {
+	client.Client
+	http.Server
+	Mux *http.ServeMux
+	Log logr.Logger
+}
+
+// Add creates a new Git Webhook Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	s := NewServer(mgr, ":8080")
+	return mgr.Add(s)
+}
+
+func NewServer(mgr manager.Manager, addr string) *Server {
+	s := &Server{
+		Client: mgr.GetClient(),
+		Server: http.Server{
+			Addr: addr,
+		},
+		Log: logf.Log.WithName("webhook-server"),
+	}
+	s.Mux = http.NewServeMux()
+	s.Server.Handler = s.Mux
+
+	s.Mux.HandleFunc("/github", s.githubWebhook)
+
+	return s
+}
+
+func (s *Server) Start(stop <-chan struct{}) error {
+	s.Log.Info("webhook server is listening", "address", s.Addr)
+
+	if err := s.ListenAndServe(); err != http.ErrServerClosed {
+		return err
+	}
+
+	<-stop
+	return nil
+}
+
+func (s *Server) githubWebhook(w http.ResponseWriter, r *http.Request) {
+	ghClient := github.NewDefault()
+
+	secret := func(webhook scm.Webhook) (secret string, err error) {
+		return os.Getenv("WEBHOOK_SECRET"), nil
+	}
+
+	webhook, err := ghClient.Webhooks.Parse(r, secret)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	switch event := webhook.(type) {
+	case *scm.PushHook:
+		if strings.HasPrefix(event.Ref, "refs/heads/") {
+			repo, branch, ref := event.Repo.Clone, strings.TrimPrefix(event.Ref, "refs/heads/"), event.After
+			s.Log.Info("received push event",
+				"repo", repo,
+				"branch", branch,
+				"ref", ref,
+			)
+			s.updateRefWrapper(repo, branch, ref)
+		}
+	case *scm.TagHook:
+	case *scm.BranchHook:
+	case *scm.IssueHook:
+	case *scm.IssueCommentHook:
+	case *scm.PullRequestHook:
+	case *scm.PullRequestCommentHook:
+	case *scm.ReviewCommentHook:
+	}
+}
+
+func (s *Server) updateRefWrapper(repo, branch, ref string) {
+	err := s.updateRef(repo, branch, ref)
+	if err != nil {
+		s.Log.Error(err, "received error while updating wordpress refs")
+	}
+}
+
+func (s *Server) updateRef(repo, branch, ref string) error {
+	wpList := v1alpha1.WordpressList{}
+
+	err := s.Client.List(context.TODO(), &client.ListOptions{}, &wpList)
```

Comment:
You should create an indexer like the one in https://github.com/presslabs/dashboard/blob/38b195b0dea4ee6267b17ea71eb82d3342119115/pkg/apiserver/controller/organization/organization.go#L48

You should index for `git.ref` and `git.repo`. You should also normalize the git repo so that:
* `git@github.com:OWNER/REPO` becomes `github.com/OWNER/REPO`
* `git@github.com:OWNER/REPO.git` becomes `github.com/OWNER/REPO`
* `https://github.com/OWNER/REPO` becomes `github.com/OWNER/REPO`
* `https://github.com/OWNER/REPO.git` becomes `github.com/OWNER/REPO`